packages/flutter_hooks/lib/src/listenable.dart

```diff
@@ -173,7 +171,8 @@ class _OnListenableChangeHookState
   @override
   void didUpdateHook(_OnListenableChangeHook oldHook) {
     super.didUpdateHook(oldHook);
-    if (hook.listenable != oldHook.listenable) {
+    if (hook.listenable != oldHook.listenable ||
```

Comment:
That's not very efficient. It'd usually add/remove a listener whenever the build method updates.
We should have a `listener` method on the Hook class.

Of course we'd need a test for that too :)

Reply:
Hmm you're right, I never thought about that, most people are going to use it with anonymous functions. Time to refactor all my codebases that use callbacks in hooks as well 😅

Reply:
I've updated it, I'm not quite sure if this is what you meant, but it seems to work

Reply:
The issue isn't the anonymous function. It's your didUpdateHook implementation.

Like I said, you could have:

```dart
class A extends Hook {
  void listener() {
    hook.listener();
  }

  initHook() {
    listenable.addListener(listener);
  }
  dispose() {
    listenable.removeListener(listener);
  }
}
```

Reply:
Thanks for the feedback, however, I honestly can't quite make sense of what you're saying.

Both your previous comment, as well as this one refer to having a `listener` method on the hook class, however your pseudo-example calls `hook.listener()` from inside a `Hook` not a `HookState`, which must be a mistake. 

I'm therefore assuming you meant to write `extends HookState`. It makes sense to optimize `listener` changes like this, which is what I did on the most recent commits, but I'm failing to see how this would update whenever the Hook's `listenable` changes?

Reply:
Tests are still passing, aren't they? ;)

`this.hook` was mutated when `didUpdateHook` was called. So the `HookState.listener` method points to the new callback.

Reply:
I think there's a misunderstanding. If you agree with the latest implementation on this branch, we're on the same page. I was under the assumption that you want to get rid of the entire `didUpdateHook` method as well. That's the part that doesn't make sense to me.

Reply:
If the only parts where the listener is added and removed are `initHook` and `dispose` respectively, which is what I thought you were pointing at, tests don't pass anymore.